File: okhttp-tests/src/test/java/com/squareup/okhttp/DelegatingSSLSocket.java
/*
 * Copyright 2014 Square Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.squareup.okhttp;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.InetAddress;
import java.net.SocketAddress;
import java.net.SocketException;
import java.nio.channels.SocketChannel;
import javax.net.ssl.HandshakeCompletedListener;
import javax.net.ssl.SSLParameters;
import javax.net.ssl.SSLSession;
import javax.net.ssl.SSLSocket;

/**
 * An {@link javax.net.ssl.SSLSocket} that delegates all calls.
 */
public abstract class DelegatingSSLSocket extends SSLSocket {
  protected final SSLSocket delegate;

  public DelegatingSSLSocket(SSLSocket delegate) {
    this.delegate = delegate;
  }

  @Override public void shutdownInput() throws IOException {
    delegate.shutdownInput();
  }

  @Override public void shutdownOutput() throws IOException {
    delegate.shutdownOutput();
  }

  @Override public String[] getSupportedCipherSuites() {
    return delegate.getSupportedCipherSuites();
  }

  @Override public String[] getEnabledCipherSuites() {
    return delegate.getEnabledCipherSuites();
  }

  @Override public void setEnabledCipherSuites(String[] suites) {
    delegate.setEnabledCipherSuites(suites);
  }

  @Override public String[] getSupportedProtocols() {
    return delegate.getSupportedProtocols();
  }

  @Override public String[] getEnabledProtocols() {
    return delegate.getEnabledProtocols();
  }

  @Override public void setEnabledProtocols(String[] protocols) {
    delegate.setEnabledProtocols(protocols);
  }

  @Override public SSLSession getSession() {
    return delegate.getSession();
  }

  @Override public void addHandshakeCompletedListener(HandshakeCompletedListener listener) {
    delegate.addHandshakeCompletedListener(listener);
  }

  @Override public void removeHandshakeCompletedListener(HandshakeCompletedListener listener) {
    delegate.removeHandshakeCompletedListener(listener);
  }

  @Override public void startHandshake() throws IOException {
    delegate.startHandshake();
  }

  @Override public void setUseClientMode(boolean mode) {
    delegate.setUseClientMode(mode);
  }

  @Override public boolean getUseClientMode() {
    return delegate.getUseClientMode();
  }

  @Override public void setNeedClientAuth(boolean need) {
    delegate.setNeedClientAuth(need);
  }

  @Override public void setWantClientAuth(boolean want) {
    delegate.setWantClientAuth(want);
  }

  @Override public boolean getNeedClientAuth() {
    return delegate.getNeedClientAuth();
  }

  @Override public boolean getWantClientAuth() {
    return delegate.getWantClientAuth();
  }

  @Override public void setEnableSessionCreation(boolean flag) {
    delegate.setEnableSessionCreation(flag);
  }

  @Override public boolean getEnableSessionCreation() {
    return delegate.getEnableSessionCreation();
  }

  @Override public SSLParameters getSSLParameters() {
    return delegate.getSSLParameters();
  }

  @Override public void setSSLParameters(SSLParameters p) {
    delegate.setSSLParameters(p);
  }

  @Override public void close() throws IOException {
    delegate.close();
  }

  @Override public InetAddress getInetAddress() {
    return delegate.getInetAddress();
  }

  @Override public InputStream getInputStream() throws IOException {
    return delegate.getInputStream();
  }

  @Override public boolean getKeepAlive() throws SocketException {
    return delegate.getKeepAlive();
  }

  @Override public InetAddress getLocalAddress() {
    return delegate.getLocalAddress();
  }

  @Override public int getLocalPort() {
    return delegate.getLocalPort();
  }

  @Override public OutputStream getOutputStream() throws IOException {
    return delegate.getOutputStream();
  }

  @Override public int getPort() {
    return delegate.getPort();
  }

  @Override public int getSoLinger() throws SocketException {
    return delegate.getSoLinger();
  }

  @Override public int getReceiveBufferSize() throws SocketException {
    return delegate.getReceiveBufferSize();
  }

  @Override public int getSendBufferSize() throws SocketException {
    return delegate.getSendBufferSize();
  }

  @Override public int getSoTimeout() throws SocketException {
    return delegate.getSoTimeout();
  }

  @Override public boolean getTcpNoDelay() throws SocketException {
    return delegate.getTcpNoDelay();
  }

  @Override public void setKeepAlive(boolean keepAlive) throws SocketException {
    delegate.setKeepAlive(keepAlive);
  }

  @Override public void setSendBufferSize(int size) throws SocketException {
    delegate.setSendBufferSize(size);
  }

  @Override public void setReceiveBufferSize(int size) throws SocketException {
    delegate.setReceiveBufferSize(size);
  }

  @Override public void setSoLinger(boolean on, int timeout) throws SocketException {
    delegate.setSoLinger(on, timeout);
  }

  @Override public void setSoTimeout(int timeout) throws SocketException {
    delegate.setSoTimeout(timeout);
  }

  @Override public void setTcpNoDelay(boolean on) throws SocketException {
    delegate.setTcpNoDelay(on);
  }

  @Override public String toString() {
    return delegate.toString();
  }

  @Override public SocketAddress getLocalSocketAddress() {
    return delegate.getLocalSocketAddress();
  }

  @Override public SocketAddress getRemoteSocketAddress() {
    return delegate.getRemoteSocketAddress();
  }

  @Override public boolean isBound() {
    return delegate.isBound();
  }

  @Override public boolean isConnected() {
    return delegate.isConnected();
  }

  @Override public boolean isClosed() {
    return delegate.isClosed();
  }

  @Override public void bind(SocketAddress localAddr) throws IOException {
    delegate.bind(localAddr);
  }

  @Override public void connect(SocketAddress remoteAddr) throws IOException {
    delegate.connect(remoteAddr);
  }

  @Override public void connect(SocketAddress remoteAddr, int timeout) throws IOException {
    delegate.connect(remoteAddr, timeout);
  }

  @Override public boolean isInputShutdown() {
    return delegate.isInputShutdown();
  }

  @Override public boolean isOutputShutdown() {
    return delegate.isOutputShutdown();
  }

  @Override public void setReuseAddress(boolean reuse) throws SocketException {
    delegate.setReuseAddress(reuse);
  }

  @Override public boolean getReuseAddress() throws SocketException {
    return delegate.getReuseAddress();
  }

  @Override public void setOOBInline(boolean oobinline) throws SocketException {
    delegate.setOOBInline(oobinline);
  }

  @Override public boolean getOOBInline() throws SocketException {
    return delegate.getOOBInline();
  }

  @Override public void setTrafficClass(int value) throws SocketException {
    delegate.setTrafficClass(value);
  }

  @Override public int getTrafficClass() throws SocketException {
    return delegate.getTrafficClass();
  }

  @Override public void sendUrgentData(int value) throws IOException {
    delegate.sendUrgentData(value);
  }

  @Override public SocketChannel getChannel() {
    return delegate.getChannel();
  }

  @Override public void setPerformancePreferences(int connectionTime, int latency, int bandwidth) {
    delegate.setPerformancePreferences(connectionTime, latency, bandwidth);
  }
}

File: okhttp-testing-support/src/main/kotlin/okhttp3/DelegatingSSLSocket.kt
/*
 * Copyright 2014 Square Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package okhttp3

import java.io.IOException
import java.io.InputStream
import java.io.OutputStream
import java.net.InetAddress
import java.net.SocketAddress
import java.net.SocketException
import java.nio.channels.SocketChannel
import java.util.function.BiFunction
import javax.net.ssl.HandshakeCompletedListener
import javax.net.ssl.SSLParameters
import javax.net.ssl.SSLSession
import javax.net.ssl.SSLSocket

/** An [SSLSocket] that delegates all calls.  */
abstract class DelegatingSSLSocket(protected val delegate: SSLSocket?) : SSLSocket() {
  @Throws(IOException::class)
  override fun shutdownInput() {
    delegate!!.shutdownInput()
  }

  @Throws(IOException::class)
  override fun shutdownOutput() {
    delegate!!.shutdownOutput()
  }

  override fun getSupportedCipherSuites(): Array<String> {
    return delegate!!.supportedCipherSuites
  }

  override fun getEnabledCipherSuites(): Array<String> {
    return delegate!!.enabledCipherSuites
  }

  override fun setEnabledCipherSuites(suites: Array<String>) {
    delegate!!.enabledCipherSuites = suites
  }

  override fun getSupportedProtocols(): Array<String> {
    return delegate!!.supportedProtocols
  }

  override fun getEnabledProtocols(): Array<String> {
    return delegate!!.enabledProtocols
  }

  override fun setEnabledProtocols(protocols: Array<String>) {
    delegate!!.enabledProtocols = protocols
  }

  override fun getSession(): SSLSession {
    return delegate!!.session
  }

  override fun addHandshakeCompletedListener(listener: HandshakeCompletedListener) {
    delegate!!.addHandshakeCompletedListener(listener)
  }

  override fun removeHandshakeCompletedListener(listener: HandshakeCompletedListener) {
    delegate!!.removeHandshakeCompletedListener(listener)
  }

  @Throws(IOException::class)
  override fun startHandshake() {
    delegate!!.startHandshake()
  }

  override fun setUseClientMode(mode: Boolean) {
    delegate!!.useClientMode = mode
  }

  override fun getUseClientMode(): Boolean {
    return delegate!!.useClientMode
  }

  override fun setNeedClientAuth(need: Boolean) {
    delegate!!.needClientAuth = need
  }

  override fun setWantClientAuth(want: Boolean) {
    delegate!!.wantClientAuth = want
  }

  override fun getNeedClientAuth(): Boolean {
    return delegate!!.needClientAuth
  }

  override fun getWantClientAuth(): Boolean {
    return delegate!!.wantClientAuth
  }

  override fun setEnableSessionCreation(flag: Boolean) {
    delegate!!.enableSessionCreation = flag
  }

  override fun getEnableSessionCreation(): Boolean {
    return delegate!!.enableSessionCreation
  }

  override fun getSSLParameters(): SSLParameters {
    return delegate!!.sslParameters
  }

  override fun setSSLParameters(p: SSLParameters) {
    delegate!!.sslParameters = p
  }

  @Throws(IOException::class)
  override fun close() {
    delegate!!.close()
  }

  override fun getInetAddress(): InetAddress {
    return delegate!!.inetAddress
  }

  @Throws(IOException::class)
  override fun getInputStream(): InputStream {
    return delegate!!.inputStream
  }

  @Throws(SocketException::class)
  override fun getKeepAlive(): Boolean {
    return delegate!!.keepAlive
  }

  override fun getLocalAddress(): InetAddress {
    return delegate!!.localAddress
  }

  override fun getLocalPort(): Int {
    return delegate!!.localPort
  }

  @Throws(IOException::class)
  override fun getOutputStream(): OutputStream {
    return delegate!!.outputStream
  }

  override fun getPort(): Int {
    return delegate!!.port
  }

  @Throws(SocketException::class)
  override fun getSoLinger(): Int {
    return delegate!!.soLinger
  }

  @Throws(SocketException::class)
  override fun getReceiveBufferSize(): Int {
    return delegate!!.receiveBufferSize
  }

  @Throws(SocketException::class)
  override fun getSendBufferSize(): Int {
    return delegate!!.sendBufferSize
  }

  @Throws(SocketException::class)
  override fun getSoTimeout(): Int {
    return delegate!!.soTimeout
  }

  @Throws(SocketException::class)
  override fun getTcpNoDelay(): Boolean {
    return delegate!!.tcpNoDelay
  }

  @Throws(SocketException::class)
  override fun setKeepAlive(keepAlive: Boolean) {
    delegate!!.keepAlive = keepAlive
  }

  @Throws(SocketException::class)
  override fun setSendBufferSize(size: Int) {
    delegate!!.sendBufferSize = size
  }

  @Throws(SocketException::class)
  override fun setReceiveBufferSize(size: Int) {
    delegate!!.receiveBufferSize = size
  }

  @Throws(SocketException::class)
  override fun setSoLinger(
    on: Boolean,
    timeout: Int,
  ) {
    delegate!!.setSoLinger(on, timeout)
  }

  @Throws(SocketException::class)
  override fun setSoTimeout(timeout: Int) {
    delegate!!.soTimeout = timeout
  }

  @Throws(SocketException::class)
  override fun setTcpNoDelay(on: Boolean) {
    delegate!!.tcpNoDelay = on
  }

  override fun toString(): String {
    return delegate!!.toString()
  }

  override fun getLocalSocketAddress(): SocketAddress {
    return delegate!!.localSocketAddress
  }

  override fun getRemoteSocketAddress(): SocketAddress {
    return delegate!!.remoteSocketAddress
  }

  override fun isBound(): Boolean {
    return delegate!!.isBound
  }

  override fun isConnected(): Boolean {
    return delegate!!.isConnected
  }

  override fun isClosed(): Boolean {
    return delegate!!.isClosed
  }

  @Throws(IOException::class)
  override fun bind(localAddr: SocketAddress) {
    delegate!!.bind(localAddr)
  }

  @Throws(IOException::class)
  override fun connect(remoteAddr: SocketAddress) {
    delegate!!.connect(remoteAddr)
  }

  @Throws(IOException::class)
  override fun connect(
    remoteAddr: SocketAddress,
    timeout: Int,
  ) {
    delegate!!.connect(remoteAddr, timeout)
  }

  override fun isInputShutdown(): Boolean {
    return delegate!!.isInputShutdown
  }

  override fun isOutputShutdown(): Boolean {
    return delegate!!.isOutputShutdown
  }

  @Throws(SocketException::class)
  override fun setReuseAddress(reuse: Boolean) {
    delegate!!.reuseAddress = reuse
  }

  @Throws(SocketException::class)
  override fun getReuseAddress(): Boolean {
    return delegate!!.reuseAddress
  }

  @Throws(SocketException::class)
  override fun setOOBInline(oobinline: Boolean) {
    delegate!!.oobInline = oobinline
  }

  @Throws(SocketException::class)
  override fun getOOBInline(): Boolean {
    return delegate!!.oobInline
  }

  @Throws(SocketException::class)
  override fun setTrafficClass(value: Int) {
    delegate!!.trafficClass = value
  }

  @Throws(SocketException::class)
  override fun getTrafficClass(): Int {
    return delegate!!.trafficClass
  }

  @Throws(IOException::class)
  override fun sendUrgentData(value: Int) {
    delegate!!.sendUrgentData(value)
  }

  override fun getChannel(): SocketChannel {
    return delegate!!.channel
  }

  override fun getHandshakeSession(): SSLSession {
    return delegate!!.handshakeSession
  }

  override fun getApplicationProtocol(): String {
    return delegate!!.applicationProtocol
  }

  override fun getHandshakeApplicationProtocol(): String {
    return delegate!!.handshakeApplicationProtocol
  }

  override fun setHandshakeApplicationProtocolSelector(selector: BiFunction<SSLSocket, MutableList<String>, String>?) {
    delegate!!.handshakeApplicationProtocolSelector = selector
  }

  override fun getHandshakeApplicationProtocolSelector(): BiFunction<SSLSocket, MutableList<String>, String> {
    return delegate!!.handshakeApplicationProtocolSelector
  }
}
